_posts/2016-09-14-programming-model.md

```diff
@@ -0,0 +1,128 @@
+
+In GFX, when you create a texture you only provide the surface format, which dictates how many bits each channel has. Then, for the view creation, you specify how these bits are interpret (normalized integers, floats, raw ints, etc). Note: the resources are created by a `gfx::Factory` object, which may or may not be sendable/clonable, depending on the backend features. Example:
+{% highlight Rust %}
+let texture = try!(factory.create_texture(kind, levels, SHADER_RESOURCE | RENDER_TARGET, Usage::GpuOnly, None)); // providing the storage configuration and requesting the capabilities
+let resource = try!(factory.view_texture_as_shader_resource::<Srgba8>(&texture, (0, levels-1), format::Swizzle::new())); // creating an SRV for the whole mipmap range
+let target = try!(factory.view_texture_as_render_target(&texture, 0, None)); // creating an RTV for level 0
+{% endhighlight %}
+
+There are helper methods to create the texture with the views for your convenience (`create_render_target`, `create_depth_stencil`, etc). Later, the texture object itself does not participate in any rendering. It can only be used for updating the contents and reading them back. Typically, you'd ignore this object and only work with its views, given that they hold the parent object automatically.
+
+Note: applied to OpenGL, the GFX texture and RTV/DSV are mapping to the same thing (GL texture). The GL render buffer is used instead if `SHADER_RESOURCE` capability is not requested.
+
+
+## Pipeline states
+
+The general trend in graphics hardware is to do more and more with programmable shaders. This has been the case in consumer cards for a long time, since the first assembly shaders came into play. Thus, the states that used to map directly to switches in fixed-function hardware (like vertex attribute fetches) can now be compiled into a small chunk of shader code and executed somewhere in-between your shader code. This compile step and hookup can introduce lags and requires some anticipation by the driver, making those drivers rather bloated and full of quirks (part of the reason AMD/NV supply game profiles in the drivers). Current-gen APIs largely address this with pipeline states.
+
+A pipeline state objects (PSO) encapsulates all the dependent states of your shader program (with the program itself). You declare in advance what kind of vertex attributes will come as input, what render targets to draw to, with what blending modes, what textures/samplers to use, etc. This allows the driver to compile the most efficient code for the set of states. Batching becomes trivial - just send multiple draw calls with the same PSO to get the maximum performance. The downside here is the inconvenience and restrictive nature of these forward state declarations.
+
+Our PSOs are defined with macros like this:
```

Comment:
> Our PSOs are defined with macros like this:

Maybe an analogy to Rust macros for those not familiar, or link to the previous blog post introducing PSOs (though a bit outdated)